lease/lessor_bench_test.go

```diff
@@ -15,112 +15,202 @@
-
-func benchmarkLessorFindExpired(size int, b *testing.B) {
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	le.Promote(0)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+func BenchmarkLessorGrant1000(b *testing.B)   { benchmarkLessorGrant(1000, b) }
+func BenchmarkLessorGrant100000(b *testing.B) { benchmarkLessorGrant(100000, b) }
+
+func BenchmarkLessorRevoke1000(b *testing.B)   { benchmarkLessorRevoke(1000, b) }
+func BenchmarkLessorRevoke100000(b *testing.B) { benchmarkLessorRevoke(100000, b) }
+
+func BenchmarkLessorRenew1000(b *testing.B)   { benchmarkLessorRenew(1000, b) }
+func BenchmarkLessorRenew100000(b *testing.B) { benchmarkLessorRenew(100000, b) }
+
+// Use findExpired10000 replace findExpired1000, which takes too long.
+func BenchmarkLessorFindExpired10000(b *testing.B)  { benchmarkLessorFindExpired(10000, b) }
+func BenchmarkLessorFindExpired100000(b *testing.B) { benchmarkLessorFindExpired(100000, b) }
+
+func init() {
+	rand.Seed(time.Now().UTC().UnixNano())
+}
+
+func randomTTL(n int, min, max int64) (out []int64) {
+	for i := 0; i < n; i++ {
+		out = append(out, rand.Int63n(max-min)+min)
 	}
-	le.mu.Lock() //Stop the findExpiredLeases call in the runloop
+	return out
+}
+
+// demote lessor from being the primary, but don't change any lease's expiry
+func demote(le *lessor) {
+	le.mu.Lock()
 	defer le.mu.Unlock()
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		le.findExpiredLeases(1000)
-	}
+	close(le.demotec)
+	le.demotec = nil
 }
 
-func benchmarkLessorGrant(size int, b *testing.B) {
+// return new lessor and tearDown to release resource
+func setUp() (le *lessor, tearDown func()) {
 	lg := zap.NewNop()
 	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
-	}
-	b.ResetTimer()
-	for i := 0; i < b.N; i++ {
-		le.Grant(LeaseID(i+size), int64(100+i+size))
+	// MinLeaseTTL is negative, so we can grant expired lease in benchmark.
+	le = newLessor(lg, be, LessorConfig{MinLeaseTTL: -100})
+	le.SetRangeDeleter(func() TxnDelete {
+		ftd := &FakeTxnDelete{be.BatchTx()}
+		ftd.Lock()
+		return ftd
+	})
+	le.Promote(0)
+
+	return le, func() {
+		le.Stop()
+		be.Close()
+		os.Remove(tmpPath)
 	}
 }
 
-func benchmarkLessorRevoke(size int, b *testing.B) {
-	lg := zap.NewNop()
-	be, tmpPath := backend.NewDefaultTmpBackend()
-	le := newLessor(lg, be, LessorConfig{MinLeaseTTL: minLeaseTTL})
-	defer le.Stop()
-	defer cleanup(be, tmpPath)
-	for i := 0; i < size; i++ {
-		le.Grant(LeaseID(i), int64(100+i))
+func benchmarkLessorGrant(benchSize int, b *testing.B) {
+	// avoid lease expire when benchmark
+	ttls := randomTTL(benchSize, 10, 100)
```

Comment:
How are the min and max chosen? Do they always guarantee that no leases expire during the benchmark?

Reply:
The best:
we chosen suitable min and max, make the repeat probability of ttls meet the prod situation.
This is too hard. So i choose two random values.

10 second can avoid expire under 99.99%.
And under other 0.01%, the benchmark will run too long.

Reply:
Sorry I do not fully understand. It seems that the safe value of min TTL should depend on benchSize and how fast the test machine runs? Maybe 10s is large enough to cover all cases? Can we use really large values to make it obvious that the leases won't expire during the test?

Reply:
Ok, I change min ttl to 1000 second.